tests/test_algorithm/test_ioscm.py

```diff
@@ -77,10 +77,384 @@ def test_get_consolidated_district(self) -> None:
+                    )
+
+    def test_get_consolidated_district_preserves_district_membership(self) -> None:
+        """Test that function correctly identifies district membership. (i.e. which nodes belong to which district).
+
+        Graph: simple_cyclic_graph_1
+        Structure: R -> X -> W -> Z -> X (cycle), W -> Y
+        Districts: {R}, {X, W, Z}, {Y}
+
+
+        Test Case 1: Nodes from different districts
+        Query: {R, X} -> Expected Output: {frozenset({R}), frozenset({X, W, Z})}
+        Query: {R, Y} -> Expected Output: {frozenset({R}), frozenset({Y})}
+        Query: {X, Y} -> Expected Output: {frozenset({X, W, Z}), frozenset({Y})}
+
+        Test Case 2: Nodes from same district
+        Query: {X, W} -> Expected Output: {X, W, Z}
+        Query: {W, Z} -> Expected Output: {X, W, Z}
+        Query: {X, Z} -> Expected Output: {X, W, Z}
+        """
+        # simple_cyclic_graph has three districts: {R}, {X, W, Z}, {Y}
+
+        # Test 1: test all pairs of nodes to verify district membership for different districts
+        different_district_pairs = [
+            ({R, X}, {frozenset({R}), frozenset({X, W, Z})}),
+            ({R, Y}, {frozenset({R}), frozenset({Y})}),
+            ({X, Y}, {frozenset({X, W, Z}), frozenset({Y})}),
+        ]
+
+        for query, expected in different_district_pairs:
+            with self.subTest(query=query):
+                result = get_consolidated_district(simple_cyclic_graph_1, query)
+                self.assertSetEqual(result, expected, f"Nodes {query} are in different districts.")
+
+        # Test 2: test pairs from same district
+        same_district_pairs = [
+            ({X, W}, {X, W, Z}),
+            ({W, Z}, {X, W, Z}),
+            ({X, Z}, {X, W, Z}),
+        ]
+
+        for query, expected in same_district_pairs:
+            with self.subTest(query=query):
+                result = get_consolidated_district(simple_cyclic_graph_1, query)
+                self.assertSetEqual(result, expected, f"Nodes {query} are in the same district.")
+
+    def test_get_consolidated_district_return_type_consistency(self) -> None:
+        """Test that return type is predictable based on district count.
+
+        Rule:
+        - 1 district -> set[Variable]
+        - 2+ districts -> set[frozenset[Variable]]
+
+        Test Cases for Single district:
+
+        Graph 1: simple_cyclic_graph_1
+        1. Query: {X} -> Output: set of Variables {X, W, Z}
+        2. Query: {X, W, Z} -> Output: set of Variables {X, W, Z} (all from the same district)
+
+
+        Graph 2: simple_cyclic_graph_2
+        3. Query: {R, X} -> Output: set of Variables {R, X, W, Z}
+        4. Query: {R, X, W, Z} -> Output: set of Variables {R, X, W, Z} (all from the same district)
+
+        Test Cases for Multiple Districts:
+
+        Graph 1 (simple_cyclic_graph_1):
+        5. Query {X, R} → Returns {frozenset({X,W,Z}), frozenset({R})} (2 districts)

```

Comment:
this kind of documentation isn't great because it simply duplicates what's written in the code. I would say just remove it.